scio-elasticsearch/src/main/java/com/spotify/scio/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,265 @@
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      public Bound<T> withFunction(SerializableFunction<T, IndexRequest> function) {
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      public Bound<T> withNumOfShard(Long numOfShard) {
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      @Override
+      public PDone expand(final PCollection<T> input) {
+
+        if (clusterName == null) {
+          throw new IllegalStateException(
+              "need to set clustername of ElasticsearchIO.Write transform");
+        }
+
+        if (servers == null) {
+          throw new IllegalStateException(
+              "need to set clustername of ElasticsearchIO.Write transform");
+        }
+
+        if (function == null) {
+          throw new IllegalStateException(
+              "need to set SerializableFunction<T, IndexRequest> of ElasticsearchIO.Write transform");
+        }
+
+        if (numOfShard == null) {
+          throw new IllegalStateException(
+              "need to set numOfShard of ElasticsearchIO.Write transform");
+        }
+        input
+            .apply("Assign To Shard", ParDo.of(new AssignToShard<>(numOfShard)))
+            .apply("Re-Window to Global Window", Window.<KV<Long, T>>into(new GlobalWindows())
+                       .triggering(Repeatedly.forever(
+                           AfterProcessingTime
+                               .pastFirstElementInPane()
+                               .plusDelayOf(javaToJoda(flushInterval))))
+                       .discardingFiredPanes())
+            .apply(GroupByKey.create())
+            .apply("Write to Elasticesarch",
+                   ParDo.of(new ElasticsearchWriter(clusterName, servers, function)));
+        return PDone.in(input.getPipeline());
+      }
+      private org.joda.time.Duration javaToJoda(final Duration duration) {
+        return duration == null ? null : org.joda.time.Duration.millis(duration.toMillis());
+      }
+    }
+
+    private static class AssignToShard<T> extends DoFn<T, KV<Long, T>> {
+      private final Long numOfShard;
+
+      public AssignToShard(Long numOfShard) {
+        this.numOfShard = numOfShard;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        // assign this element to a random shard
+        final long shard = ThreadLocalRandom.current().nextLong(numOfShard);
+        c.output(KV.of(shard, c.element()));
+      }
+    }
+
+    private static class ElasticsearchWriter<T> extends DoFn<KV<Long, Iterable<T>>, Void> {
+      private final Logger LOG = LoggerFactory.getLogger(ElasticsearchWriter.class);
+      private final ClientSupplier clientSupplier;
+      private final SerializableFunction<T, IndexRequest> function;
+
+      public ElasticsearchWriter(String clusterName,
```

Comment:
@regadas Following Beam/Dataflow convention of providing factory with individual fields instead of accepting object with all fields included.